### PR TITLE
Add FMC SDRAM for STM32H7x3 chip

### PR DESCRIPTION
**Pull request by yukihiratype2**
The STM32H7 chip's FMC is different from F7 or F4. H7 doesn't have BUSY register, instead, H7 requires set FMCEN on FMC_BCR1 after configured FMC registers.

### CODE REVIEW

File: arch/arm/src/stm32h7/Kconfig [hunk 1]
```diff
@@ -41,6 +41,7 @@ config STM32H7_STM32H7X3XX
 	select ARMV7M_HAVE_ITCM
 	select ARMV7M_HAVE_DTCM
 	select STM32H7_HAVE_ETHERNET
+	select STM32H7_HAVE_FMC
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/Kconfig]: Hi, from what I see the whole STM32H7 family has support for FMC, so probably we can enable this feature also for STM32H7_STM32H7X7XX

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+ * Name: stm32_fmc_sdram_write_protect
+ *
+ * Description:
+ *   Enable/Disable writes to an SDRAM.
+ *
+ ****************************************************************************/
+
+void stm32_fmc_sdram_write_protect(int bank, bool state)
+{
+  uint32_t val, sdcr;
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_fmc.c]: Nuttx Coding Standard requires one declaration per line:
http://nuttx.org/Documentation/NuttXCCodingStandard.html#onedatperline

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+ * Name: stm32_fmc_sdram_set_timing
+ *
+ * Description:
+ *   Set the SDRAM timing parameters.
+ *
+ ****************************************************************************/
+
+void stm32_fmc_sdram_set_timing(int bank, uint32_t timing)
+{
+  uint32_t val, sdtr;
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_fmc.c]: Nuttx Coding Standard requires one declaration per line:
http://nuttx.org/Documentation/NuttXCCodingStandard.html#onedatperline

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+ * Name: stm32_fmc_sdram_set_control
+ *
+ * Description:
+ *   Set the SDRAM control parameters.
+ *
+ ****************************************************************************/
+
+void stm32_fmc_sdram_set_control(int bank, uint32_t ctrl)
+{
+  uint32_t val, sdcr;
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_fmc.c]: Nuttx Coding Standard requires one declaration per line:
http://nuttx.org/Documentation/NuttXCCodingStandard.html#onedatperline

File: arch/arm/src/stm32h7/stm32_fmc.h [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************************/
+
+#ifndef __ARCH_ARM_STC_STM32H7_STM32_FMC_H
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_fmc.h]: should be __ARCH_ARM_SRC_STM32H7_STM32_FMC_H not __ARCH_ARM_STC_STM32H7_STM32_FMC_H

File: arch/arm/src/stm32h7/hardware/stm32_fmc.h [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+#define FMC_BCR_BURSTEN          (1 << 8)   /* Burst enable bit */
+#define FMC_BCR_WAITPOL          (1 << 9)   /* Wait signal polarity bit */
+#define FMC_BCR_WRAPMOD          (1 << 10)  /* Wrapped burst mode support */
+#define FMC_BCR_WAITCFG          (1 << 11)  /* Wait timing configuration */
+#define FMC_BCR_WREN             (1 << 12)  /* Write enable bit */
+#define FMC_BCR_WAITEN           (1 << 13)  /* Wait enable bit */
+#define FMC_BCR_EXTMOD           (1 << 14)  /* Extended mode enable */
+#define FMC_BCR_ASYNCWAIT        (1 << 15)  /* Wait signal during asynchronous transfers */
+#define FMC_BCR_CBURSTRW         (1 << 19)  /* Write burst enable */
+#define FMC_BCR_BMAP_SHIFT    (24)
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/hardware/stm32_fmc.h]: bad alignment

File: arch/arm/src/stm32h7/stm32_fmc.h (hunk 5 above)
Comment (yukihiratype2) [reply arch/arm/src/stm32h7/stm32_fmc.h]: wow, that's close

File: arch/arm/src/stm32h7/Make.defs [hunk 7]
```diff
@@ -141,6 +141,10 @@ ifeq ($(CONFIG_STM32H7_DMA),y)
 CHIP_CSRCS += stm32_dma.c
 endif
 
+ifeq ($(CONFIG_STM32H7_FMC),y)
+CHIP_CSRCS += stm32_fmc.c
+endifo
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/Make.defs]: typo, I think it should be endif ;)

Comment (yukihiratype2) [reply arch/arm/src/stm32h7/Make.defs]: feel shame

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 8]
```diff
@@ -0,0 +1,196 @@
+/************************************************************************************
+ * arch/arm/src/stm32h7/stm32_fmc.c
+ *
+ *   Copyright (C) 20019 Gregory Nutt. All rights reserved.
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_fmc.c]: typo, should be 2019

File: arch/arm/src/stm32h7/stm32_fmc.c [hunk 9]
```diff
@@ -0,0 +1,196 @@
+/************************************************************************************
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_fmc.c]: The length of the comments section should be identical throughout the file. 
There is no need for such long lines in this file. The same applies to stm32_fmc.h file.

I see that this looks the same in arch/stm32 - I'll fix that.

File: arch/arm/src/stm32/stm32_fmc.c [hunk 10]
```diff
@@ -1,7 +1,7 @@
 /************************************************************************************
```
Comment (raiden00pl) [inline arch/arm/src/stm32/stm32_fmc.c]: Could you also fix the length of the comment section here and in stm32_fmc.h ?

File: arch/arm/src/stm32h7/hardware/stm32_fmc.h [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+#define FMC_BCR_WAITCFG          (1 << 11)  /* Wait timing configuration */
+#define FMC_BCR_WREN             (1 << 12)  /* Write enable bit */
+#define FMC_BCR_WAITEN           (1 << 13)  /* Wait enable bit */
+#define FMC_BCR_EXTMOD           (1 << 14)  /* Extended mode enable */
+#define FMC_BCR_ASYNCWAIT        (1 << 15)  /* Wait signal during asynchronous transfers */
+#define FMC_BCR_CBURSTRW         (1 << 19)  /* Write burst enable */
+#define FMC_BCR_BMAP_SHIFT       (24)
+#  define FMC_BCR_BMAP_0               (0 << FMC_BCR_BMAP_SHIFT)
+#  define FMC_BCR_BMAP_1               (1 << FMC_BCR_BMAP_SHIFT)
+#  define FMC_BCR_BMAP_2               (2 << FMC_BCR_BMAP_SHIFT)
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/hardware/stm32_fmc.h]: bad alignment here too.
Sorry, in previous comment I should select multiple lines for comment.

Comment (yukihiratype2) [reply arch/arm/src/stm32h7/hardware/stm32_fmc.h]: No, that's my fault.
I should check before push

File: arch/arm/include/stm32h7/chip.h [hunk 12]
```diff
@@ -167,9 +167,9 @@
 #endif
 
 #if defined(CONFIG_STM32F7_HAVE_FMC)
```
Comment (raiden00pl) [inline arch/arm/include/stm32h7/chip.h]: should be CONFIG_STM32H7_HAVE_FMC

**Pull request merged**